### ISSUE TITLE
Accessibility for Visually Impaired - Work with Orca Screen Reader

### ISSUE THREAD

**Issue by abhijithb21** [i3/i3#3393]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

Not at all working with screen reader.
## Expected Behavior

When screen reader is active visually impaired persons shall be able to listen to the contents in the i3bar. They should be able to navigate through the items and the screen reader announces the item on which the focus is on. In DEs like GNOME or MATE people use Alt + Super + S key to toggle screen reader. Then they use some shortcut specified by the DE to switch focus between panels. In each panel they use arrow key to navigate. Technically what we need (IMHO) is i3bar and dmenu to communicate with the ATSPI (https://developer.gnome.org/libatspi/stable/AtspiAccessible.html).
## Environment

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 
</pre>

<pre>
</pre>

<pre>
Logfile URL:
</pre>

<pre>
- Linux Distribution & Version:
- Are you using a compositor (e.g., xcompmgr or compton):
</pre>

**Comment by Airblader** [comment]
Thanks for reporting this! I think a11y is a topic that i3 currently doesn't cover in a meaningful way, but I'm definitely open to changing that. It would be great if we could gain some insight on the details of what this particular change would require in i3bar (or maybe some type of prototype).

Small note:

> dmenu to communicate with the ATSPI 

dmenu isn't developed by i3.

**Comment by abhijithb21** [comment]
I do not have much idea regarding that but I am subscribed to Orca mailing list. Since you are interested in implementing a11y I shall share the link to this issue in the list. It is very active and I am sure someone will respond soon.

**Comment by abhijithb21** [comment]
Untill now, all I could get from the list is that a developer StormDragon fixed a11y for Ratpoison(http://www.nongnu.org/ratpoison/), another tiling minimalistic window manager. Let me see if I can get in touch with him. 
Meanwhile, GNOME has a wiki on a11y - https://wiki.gnome.org/Accessibility.

**Comment by orestisf1993** [comment]
> developer StormDragon fixed a11y for Ratpoison

https://github.com/stormdragon2976/strychnine

But I think it doesn't modify ratpoison's status bar to communicate with atspi.

I think we'd need to play with Glib to implement this.

**Comment by pvagner** [comment]
I am also one of the interested users thus my development related suggestions are just observations of behaviour rather than true development experience.
I'd say accessibility should be implemented gradually into projects such as i3. Initial inspiration should be gathered by looking at window managers such as mate/marco or gnome/metacity. Initially it would be huge step forward if icons signalling window changes (such as when pressing alt+tab) were communicated to assistive tools via at-spi.
If i3 has some internal widget like components then perhaps it might be possible to look into mapping these widgets to ATK roles and slowly implementing relevant ATK interfaces for those one at a time. For static UI components such as buttons, icons this may not be too difficult I hope.
For the rest better knowledge of i3 inner workings and accessibility experience are essential I guess.

**Comment by abhijithb21** [comment]
I guess most of the i3bar is text. Icons come into play when an application is running on background and it stays minimized at i3bar. Eg: Kazam screen recorder.

**Comment by RichMorin** [comment]
This is kind of a random thought, but...  All of the screen readers already know how to handle HTML-based content. How about extending (say) i3 to act as a personal web server?  That is, map all of its layout features into HTML. How hard/useful would this be?

**Comment by RichMorin** [comment]
Here is some information from Mike Ray, on the mailing list:

> I have tried i3. I installed it on a Dell D630 laptop running first
Debian and then Arch.

> It is inaccessible in its current form because both the task bar and
the chosen menu are not accessible and are totally silent to Orca.

> In addition, somehow window sizes are not being correctly sent to
at-spi2 or something because either this subsystem or Orca itself spits
out a load of error messages with some applications, including the scite
editor.

> And Firefox is totally silent, where it is not on other desktops.

> This is a pity because i3 is very lean and would make a good choice for
platforms such as the Raspberry Pi.

> But at the very least there would need to be a rewrite of the task bar,
or a totally new one, perhaps in wxWidgets.

> There is something called the nag bar, which is the same deal. Inaccessible.

I agree about the RasPi, but my primary interest is in use on repurposed mobile phones, running (say) postmarketOS. Anyway, I asked Mike for more details (What changes would i3 need, in order to let it solve the problems you list?  Are these tweaks or would they require major surgery?) and he replied:

> Forgive me if this is wrong, I don't think it is, but it is some time
since I took a good look at the i3 source code.

> At the bare minimum it would need a new task bar, a new 'nag bar'
whatever that is, and a new menu for launching applications.

> The task bar and the nag bar are two instances of the same code if I
remember rightly, and they are totally inaccessible, and I don't think
they are even GTK if I remember rightly.

> The menu currently used to launch applications is third party, IOW not
part of i3, per se.

> Orca works via something called the 'at-spi', which is the 'access
technology service provider interface'.

> An accessible (GTK) application informs 'at-spi' of the things that
matter to accessibility, like the type of a control, it's role etc.  For
example it tells at-spi that the control which has just gained focus is
a combo box, or a button etc.

> Orca then becomes a client of the at-spi and can read this stuff.

> I am not totally familiar with how applications interface to at-spi, but
there is some kind of problem in i3, because, as I suggested before,
some applications are silent and some are not. And others produce
profuse error messages saying something about the window size not being
correctly reported, or requested.

**Comment by dstpierre** [comment]
Hi, I'm a blind person using i3 for many years and I'm converging to start using screen reader as my visual disease is a degenerating one and I'm kind of preparing for the worst, which is not being able to read on the screen.

Long story short, I kind of love my setup with Arch and i3 and was wondering if I could help in any way here.

I'll admit that my last C program is > 20 years old, I've been building backend for web apps in all sorts of languages (Go, Elixir, C#, etc), but C is kind of far.

I'm ready to give a hand here with this since I'd really like to be able to keep my setup.

Is that even possible, Is it realistic to say that an external dev could jump into the i3 source code and help with this functionality?

How would one even start?

In any case let me know what I can do.

Thanks

**Comment by Airblader** [comment]
What about i3 as a window manager would be the most important to make more accessible? I think that's a question we need to answer.

I'm not very concerned about i3bar. Our architecture makes it very easy to exchange it for any third party tool, or rewrite it against the IPC. To be honest, the latter might well be easier than trying to change i3bar directly.

However, the actual window manager is a different topic, and for that I lack the knowledge on what we'd have to achieve, so input would be welcome here.

**Comment by dstpierre** [comment]
I'm not even sure if it's i3 related or not. But here's my experience so far:

I've installed Orca and it's kind of reading things for its own config UI `orca -s` 

But nothing else seems to be read on, for instance Chromium, xterm, vscode. 

I will have to test Orca in a Gnome setup first to make sure it's working there.

But I was in the impression that i3 was the cause of not getting the screen read at all via Orca, but maybe I'm not very clear on what's the under-the-hood difference between i3 and gnome when it comes to application emiting the accessible info for the screen reader to pick.

I'm not talking about i3bar or anything, this could easily be made accessible in the config file worst case, using espeak directly, say when switching desktop.

My issue is more like all apps I'm using, none are working with the screen reader.

Maybe the first thing to understand is if it's i3 related at all, but Orca  seems to be working as-is on Gnome based installation.

**Comment by Airblader** [comment]
I see, thanks. Unfortunately I don't really know what these tools need for that, though my feeling is that as far as the client windows go, i3 should not be involved (apart from maybe things like geometry of the window if that's relevant).

**Comment by raingloom** [comment]
@dstpierre @Airblader AFAIK at the very least i3wm would need to tell the screen reader which window is selected.
I tried screen reader support only in Gnome so far, geometry doesn't seem to be relevant at all.

**Comment by LeeAlephium** [comment]
+1, today I went to went use Orca to make sure the application we are developing is accessible and learned I can't in i3.
